exercise/static/exercise/autosave.js

```diff
@@ -0,0 +1,124 @@
+    this.autoSaveUrl = null;
+    this.autoSaveIndicator = null;
+    this.init();
+  }
+
+  $.extend(AplusAutoSave.prototype, {
+    /**
+     * Initializes the autosave functionality for this form.
+     */
+    init: function() {
+      const self = this;
+
+      const submitUrl = new URL(self.element.attr("action"), window.location);
+      self.autoSaveUrl = new URL("draft/", submitUrl).href;
+
+      self.autoSaveIndicator = $("<div></div>")
+        .addClass(self.settings.autoSaveIndicatorClasses)
+        .appendTo(self.element);
+
+      const timestamp = self.element.data("draft-timestamp");
+      if (timestamp) {
+        self.setIndicatorSaveDate(new Date(timestamp));
+      }
+
+      self.element.on("input", function() {
+        self.scheduleAutoSave();
+      });
+      self.element.on("submit", function() {
+        // Cancel the autosave (if it was scheduled) when an actual submission
+        // is made.
+        clearTimeout(self.autoSaveTimeoutHandle);
+        self.autoSaveIndicator.text("");
+      });
+    },
+
+    /**
+     * Schedules an autosave for a suitable time in the future.
+     */
+    scheduleAutoSave: function() {
+      const self = this;
+
+      // Check if a save has already been scheduled. Only one save can be
+      // scheduled at a time.
+      if (self.autoSaveTimeoutHandle !== null) {
+        return;
+      }
+
+      function scheduleAutoSaveInternal() {
+        self.autoSaveTimeoutHandle = setTimeout(function() {
+          self.doAutoSave();
+        }, self.settings.autoSaveInterval);
+      }
+
+      // If a save HTTP request is currently active, wait until it's done
+      // before scheduling the next save.
+      if (self.autoSaveRequest) {
+        self.autoSaveRequest.always(scheduleAutoSaveInternal);
+      } else {
+        scheduleAutoSaveInternal();
+      }
+    },
+
+    /**
+     * Saves a draft instantly. Use scheduleAutoSave instead of this to avoid
+     * unnecessarily frequent or simultaneous autosaves.
+     */
+    doAutoSave: function() {
+      const self = this;
+      self.autoSaveIndicator.text(_("Saving draft..."));
+
+      self.autoSaveRequest = $.ajax({
+        url: self.autoSaveUrl,
+        type: "POST",
+        data: new FormData(self.dom_element),
+        contentType: false,
+        processData: false,
+      }).fail(function() {
+        self.autoSaveIndicator.text(_("Failed to save draft"));
+      }).done(function() {
+        self.setIndicatorSaveDate(new Date());
+      }).always(function() {
+        // Allow new autosave requests.
+        self.autoSaveRequest = null;
+      });
+
+      // Allow new autosaves to be scheduled.
+      self.autoSaveTimeoutHandle = null;
+    },
+
+    /**
+     * Sets the save date displayed in the autosave indicator below the
+     * exercise form.
+     */
+    setIndicatorSaveDate: function(date) {
+      this.autoSaveIndicator.text(_("Draft saved") + " " + date.toLocaleString());
```

Comment:
This depends on your browser's language settings. If I set my browser's language to Finnish, it shows the date like this: "16.11.2021 klo 10.11.17". The `data-draft-timestamp` attribute is stored as milliseconds specifically so that it can be localized by the browser instead of Django:

https://github.com/skulonen/a-plus/blob/e7edbfd53912a6c475c50e16defc034a066a8aea/exercise/submission_models.py#L537-L539

Or do you think we should match the page's language instead of the browser's? In that case, I think we could use the `lang` attribute when formatting the date in Javscript.

Reply:
I think it makes sense to match the page's language. I implemented that by using the `lang` attribute and pushed the fix.

Reply:
Matching the page's language is the right way. Unfortunately, the A+ `<html lang="en">` parameter uses only short language codes without the region (`en-GB`). JS uses the American locale for `en`, which we can't use in Finland (months and days confusingly switched). Therefore, if `lang == 'en'`, we must use `en-GB` in `date.toLocaleString('en-GB')`. Just add the if statement (with an agonizing sigh).

```javascript
let date = new Date(Date.UTC(2015, 11, 31, 16, 56, 0));
console.log(date.toLocaleString('fi'));
console.log(date.toLocaleString('en')); // ugh...
console.log(date.toLocaleString('en-GB')); // excellent, this has even 24-hour clock
31.12.2015 klo 18.56.00
12/31/2015, 6:56:00 PM
31/12/2015, 18:56:00
```

Reply:
Alright, I added the if statement along with a comment that explains it.